language/llama3-405b/SUT_VLLM.py

```diff
@@ -0,0 +1,293 @@
+from pathlib import Path
+
+import mlperf_loadgen as lg
+from dataset import Dataset
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger("Llama-70B-SUT")
+
+
+class SUT:
+    def __init__(
+        self,
+        model_path=None,
+        dtype="bfloat16",
+        batch_size=None,
+        total_sample_count=24576,
+        dataset_path=None,
+        use_cached_outputs=False,
+        # Set this to True *only for test accuracy runs* in case your prior
+        # session was killed partway through
+        workers=1,
+        tensor_parallel_size=1
+    ):
+
+        self.model_path = model_path or f"Meta-Llama-3.1-405B-Instruct{'-FP8' if dtype == 'float8' else ''}"
+
+        if not batch_size:
+            batch_size = 1
+        self.batch_size = batch_size
+
+        self.dtype = dtype
+        self.tensor_parallel_size = tensor_parallel_size
+
+        if not torch.cuda.is_available():
+            assert False, "torch gpu is not available, exiting..."
+
+        self.dataset_path = dataset_path
+        self.data_object = Dataset(
+            self.model_path,
+            dataset_path=self.dataset_path,
+            total_sample_count=total_sample_count,
+            dtype=dtype
+        )
+        self.qsl = lg.ConstructQSL(
+            self.data_object.total_sample_count,
+            self.data_object.perf_count,
+            self.data_object.LoadSamplesToRam,
+            self.data_object.UnloadSamplesFromRam,
+        )
+
+        self.load_model()
+        gen_kwargs = {
+            "temperature": 1,
+            "top_p": 1,
+            "top_k": 1,
+            "seed": 42,
+            "max_tokens": 256,
+            "min_tokens": 20
+        }
+        self.sampling_params = SamplingParams(**gen_kwargs)
+        # self.sampling_params.all_stop_token_ids.add(self.model.get_tokenizer().eos_token_id)
+
+        self.num_workers = workers
+        self.worker_threads = [None] * self.num_workers
+        self.query_queue = queue.Queue()
+
+        self.use_cached_outputs = use_cached_outputs
+        self.sample_counter = 0
+        self.sample_counter_lock = threading.Lock()
+
+    def start(self):
+        # Create worker threads
+        for j in range(self.num_workers):
+            worker = threading.Thread(target=self.process_queries)
+            worker.start()
+            self.worker_threads[j] = worker
+
+    def stop(self):
+        for _ in range(self.num_workers):
+            self.query_queue.put(None)
+
+        for worker in self.worker_threads:
+            worker.join()
+
+    def process_queries(self):
+        """Processor of the queued queries. User may choose to add batching logic"""
+
+        while True:
+            qitem = self.query_queue.get()
+            if qitem is None:
+                break
+
+            query_ids = [q.index for q in qitem]
+
+            tik1 = time.time()
+
+            input_ids_tensor = [
+                self.data_object.input_ids[q.index][:256] for q in qitem]
```

Comment:
consider changing to self.data_object.input_ids[q.index][self.gen_kwargs.max_isl]